File: apps/desktop/src/lib/events.ts
// Type definitions for Backend Events
// Must match domain/events.rs

export type AppEvent = 
  | { event: 'config:changed'; payload: { key: string; value: string } }
  // | { event: 'download:progress'; payload: { id: string; progress: number } }
;

export type EventName = AppEvent['event'];

export type EventPayload<T extends EventName> = Extract<AppEvent, { event: T }>['payload'];

File: apps/desktop/src-tauri/src/domain/events.rs
use async_trait::async_trait;
use serde::Serialize;

#[derive(Debug, Serialize, Clone)]
#[serde(tag = "event", content = "payload")] // { "event": "ConfigChanged", "payload": { ... } }
pub enum DomainEvent {
    ConfigChanged { key: String, value: String },
    // Future events:
    // DownloadProgress { id: String, progress: u32 },
    // UserLoggedIn { user_id: String },
}

impl DomainEvent {
    /// Returns the channel name for Tauri (e.g., "app://event")
    pub fn name(&self) -> &'static str {
        match self {
            DomainEvent::ConfigChanged { .. } => "config:changed",
        }
    }
}

#[async_trait]
pub trait IEventPublisher: Send + Sync {
    fn publish(&self, event: DomainEvent);
}
